# PyCity Schools Analysis

* Your analysis here
---

In [157]:
# Dependencies and Setup
from pathlib import Path
import pandas as pd

# File to Load (Remember to Change These)
school_data_to_load = Path("Resources/schools_complete.csv")
student_data_to_load = Path("Resources/students_complete.csv")

# Read School and Student Data File and store into Pandas DataFrames
school_data = pd.read_csv(school_data_to_load, encoding="utf-8")
student_data = pd.read_csv(student_data_to_load, encoding="utf-8")

# Combine the data into a single dataset.  
school_data_complete = pd.merge(student_data, school_data, how="left", on=["school_name", "school_name"])
school_data_complete.head()

Student ID     student_name gender  year        school_name  reading_score  \
0           0     Paul Bradley      M     9  Huang High School             96   
1           1     Victor Smith      M    12  Huang High School             90   
2           2  Kevin Rodriguez      M    12  Huang High School             41   
3           3    Richard Scott      M    12  Huang High School             89   
4           4       Bonnie Ray      F     9  Huang High School             87   

   maths_score  School ID        type  size   budget  
0           94          0  Government  2917  1910635  
1           43          0  Government  2917  1910635  
2           76          0  Government  2917  1910635  
3           86          0  Government  2917  1910635  
4           69          0  Government  2917  1910635

## Local Government Area Summary

In [158]:
# Calculate the Totals (Schools and Students)
school_count = len(school_data_complete['school_name'].unique())
school_count

15

In [159]:
student_count = len(school_data_complete)
student_count

39170

In [160]:
# Calculate the Total Budget
total_budget = school_data_complete['budget'].unique().sum()
total_budget

24649428

In [161]:
# Calculate the Average Scores
average_maths_score = school_data_complete['maths_score'].mean()
average_maths_score

70.33819249425581

In [162]:
# Calculate the Average Scores
average_reading_score = school_data_complete['reading_score'].mean()
average_reading_score

69.98013786060761

In [163]:
passing_math_count = school_data_complete[(school_data_complete["maths_score"] >= 50)].count()["student_name"]
passing_math_percentage = passing_math_count / float(student_count) * 100
passing_math_percentage

86.07863160582077

In [164]:
passing_reading_count = school_data_complete[(school_data_complete["reading_score"] >= 50)].count()["student_name"]
passing_reading_percentage = passing_reading_count / float(student_count) * 100
passing_reading_percentage

84.42685728874139

In [165]:
# Use the following to calculate the percentage of students that passed math and reading
passing_math_reading_count = school_data_complete[(school_data_complete["maths_score"] >= 50) & (school_data_complete["reading_score"] >= 50)].count()["student_name"]
overall_passing_rate = passing_math_reading_count /  float(student_count) * 100
overall_passing_rate

72.80827163645647

In [166]:
# Convert to DataFrame
area_summary = pd.DataFrame({"Total Schools": school_count,
                             "Total Students": student_count,
                              "Total Budget": total_budget,
                              "Average Maths Score": average_maths_score,
                              "Average Reading Score": average_reading_score,
                              "% Passing Maths": passing_math_percentage,
                              "% Passing Reading": passing_reading_percentage,
                              "% Overall Passing": overall_passing_rate
                                }, index=[0]) 

# Copy of district_summary in case I need to manipulate data later/so I can format it now
area_summary_copy = area_summary.copy()
 
# Formatting
area_summary["Total Students"] = area_summary["Total Students"].map("{:,}".format)
area_summary["Total Budget"] = area_summary["Total Budget"].map("${:,.2f}".format)

# Display the DataFrame
area_summary

Total Schools Total Students    Total Budget  Average Maths Score  \
0             15         39,170  $24,649,428.00            70.338192   

   Average Reading Score  % Passing Maths  % Passing Reading  \
0              69.980138        86.078632          84.426857   

   % Overall Passing  
0          72.808272

## School Summary

In [167]:
# Use the code provided to select the type per school from school_data
school_types = school_data.set_index('school_name')['type']
school_types

school_name
Huang High School         Government
Figueroa High School      Government
Shelton High School      Independent
Hernandez High School     Government
Griffin High School      Independent
Wilson High School       Independent
Cabrera High School      Independent
Bailey High School        Government
Holden High School       Independent
Pena High School         Independent
Wright High School       Independent
Rodriguez High School     Government
Johnson High School       Government
Ford High School          Government
Thomas High School       Independent
Name: type, dtype: object

In [168]:
# Average reading score
per_school_counts = school_data.set_index(["school_name"])["size"]
per_school_counts = school_data_complete["school_name"].value_counts()
per_school_counts

school_name
Bailey High School       4976
Johnson High School      4761
Hernandez High School    4635
Rodriguez High School    3999
Figueroa High School     2949
Huang High School        2917
Ford High School         2739
Wilson High School       2283
Cabrera High School      1858
Wright High School       1800
Shelton High School      1761
Thomas High School       1635
Griffin High School      1468
Pena High School          962
Holden High School        427
Name: count, dtype: int64

In [169]:
# Calculate the total school budget and per capita spending per school from school_data
per_school_budget = school_data.set_index('school_name')['budget']
per_school_capita = per_school_budget / per_school_counts
per_school_budget
per_school_capita

# Calculate the average test scores per school from school_data_complete
per_school_maths = school_data_complete.set_index('school_name').groupby(["school_name"])
per_school_maths = per_school_maths["maths_score"].mean()
per_school_reading = school_data_complete.set_index('school_name').groupby(["school_name"])
per_school_reading = per_school_reading["reading_score"].mean()

In [170]:
# Get the students who passed maths and passed reading by creating separate filtered DataFrames from school_data_complete.
school_passing_maths = school_data_complete.loc[school_data_complete['maths_score'] >= 50].groupby(['school_name']).size()
school_passing_reading = school_data_complete.loc[school_data_complete['reading_score'] >= 50].groupby(['school_name']).size()

# Get the students who passed both reading and maths in a separate DataFrame from school_data_complete.
passing_maths_and_reading = school_data_complete[(school_data_complete["reading_score"] >= 50) & (school_data_complete["maths_score"] >= 50)]
school_students_passing_maths_and_reading = passing_maths_and_reading.groupby(["school_name"]).size()
school_students_passing_maths_and_reading

school_name
Bailey High School       3985
Cabrera High School      1501
Figueroa High School     1995
Ford High School         1848
Griffin High School      1194
Hernandez High School    3076
Holden High School        337
Huang High School        1946
Johnson High School      3199
Pena High School          762
Rodriguez High School    3176
Shelton High School      1389
Thomas High School       1136
Wilson High School       1540
Wright High School       1435
dtype: int64

In [171]:
#  Calculate the Percentage Pass Rates
per_school_passing_maths = school_passing_maths / per_school_counts * 100
per_school_passing_reading = school_passing_reading / per_school_counts * 100
overall_passing_rate = school_students_passing_maths_and_reading / per_school_counts * 100

In [172]:
# Convert to DataFrame
per_school_summary = pd.DataFrame({'School Type': school_types,
                               'Total Students': per_school_counts,
                               'Total School Budget': per_school_budget,
                               'Per Student Budget': per_school_capita, 
                               'Average Maths Score': per_school_maths, 
                               'Average Reading Score': per_school_reading,
                               '% Passing Maths': per_school_passing_maths,
                               '% Passing Reading': per_school_passing_reading,
                               '% Overall Passing': overall_passing_rate                         
                               })

# Used .apply(lambda) to change school type from an array
per_school_summary['School Type'] = per_school_summary['School Type']

# Copy of school summary in case I need to manipulate data later/so I can give it cleaner formatting
per_school_summary_copy = per_school_summary.copy()

# Using Mapping to format Total Students and Total Budget to look like example
per_school_summary_copy['Total School Budget'] = per_school_summary_copy['Total School Budget'].map("${:,.2f}".format)
per_school_summary_copy['Per Student Budget'] = per_school_summary_copy['Per Student Budget'].map("${:.2f}".format)

per_school_summary_copy

School Type  Total Students Total School Budget  \
school_name                                                              
Bailey High School      Government            4976       $3,124,928.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Figueroa High School    Government            2949       $1,884,411.00   
Ford High School        Government            2739       $1,763,916.00   
Griffin High School    Independent            1468         $917,500.00   
Hernandez High School   Government            4635       $3,022,020.00   
Holden High School     Independent             427         $248,087.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Pena High School       Independent             962         $585,858.00   
Rodriguez High School   Government            3999       $2,547,363.00   
Shelton High School    Independent            1761       $1,056,600.00   
Thomas High School     Independent            1635       $1,043,130.00   
Wilson High School     Independent            2283       $1,319,574.00   
Wright High School     Independent            1800       $1,049,400.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Bailey High School               $628.00            72.352894   
Cabrera High School              $582.00            71.657158   
Figueroa High School             $639.00            68.698542   
Ford High School                 $644.00            69.091274   
Griffin High School              $625.00            71.788147   
Hernandez High School            $652.00            68.874865   
Holden High School               $581.00            72.583138   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Pena High School                 $609.00            72.088358   
Rodriguez High School            $637.00            72.047762   
Shelton High School              $600.00            72.034072   
Thomas High School               $638.00            69.581651   
Wilson High School               $578.00            69.170828   
Wright High School               $583.00            72.047222   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Bailey High School             87.379421          80.084405  
Cabrera High School            89.074273          80.785791  
Figueroa High School           82.807731          67.650051  
Ford High School               82.219788          67.469880  
Griffin High School            88.487738          81.335150  
Hernandez High School          81.877023          66.364617  
Holden High School             88.524590          78.922

## Top Performing Schools (By % Overall Passing)

In [184]:
# Sort and show top five schools
top_schools = per_school_summary_copy.sort_values(by=['% Overall Passing'],ascending=False)
top_schools.head(5)

School Type  Total Students Total School Budget  \
school_name                                                              
Griffin High School    Independent            1468         $917,500.00   
Cabrera High School    Independent            1858       $1,081,356.00   
Bailey High School      Government            4976       $3,124,928.00   
Wright High School     Independent            1800       $1,049,400.00   
Rodriguez High School   Government            3999       $2,547,363.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Griffin High School              $625.00            71.788147   
Cabrera High School              $582.00            71.657158   
Bailey High School               $628.00            72.352894   
Wright High School               $583.00            72.047222   
Rodriguez High School            $637.00            72.047762   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Griffin High School                71.245232        91.212534   
Cabrera High School                71.359526        90.850377   
Bailey High School                 71.008842        91.639871   
Wright High School                 70.969444        91.777778   
Rodriguez High School              70.935984        90.797699   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Griffin High School            88.487738          81.335150  
Cabrera High School            89.074273          80.785791  
Bailey High School             87.379421          80.084405  
Wright High School             86.666667          79.722222  
Rodriguez High School          87.396849          79.419855

In [113]:
top_schools.to_csv("Output/top_schools.csv", encoding="utf-8", index=False, header = True)

## Bottom Performing Schools (By % Overall Passing)

In [174]:
# Bottom Performing Schools (By % Overall Passing)
bottom_schools = per_school_summary_copy.sort_values(by=['% Overall Passing'],ascending=True)
bottom_schools.head(5)

School Type  Total Students Total School Budget  \
school_name                                                              
Hernandez High School   Government            4635       $3,022,020.00   
Huang High School       Government            2917       $1,910,635.00   
Johnson High School     Government            4761       $3,094,650.00   
Wilson High School     Independent            2283       $1,319,574.00   
Ford High School        Government            2739       $1,763,916.00   

                      Per Student Budget  Average Maths Score  \
school_name                                                     
Hernandez High School            $652.00            68.874865   
Huang High School                $655.00            68.935207   
Johnson High School              $650.00            68.843100   
Wilson High School               $578.00            69.170828   
Ford High School                 $644.00            69.091274   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Hernandez High School              69.186408        80.949299   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Wilson High School                 68.876916        82.785808   
Ford High School                   69.572472        82.438846   

                       % Passing Reading  % Overall Passing  
school_name                                                  
Hernandez High School          81.877023          66.364617  
Huang High School              81.453548          66.712376  
Johnson High School            81.978576          67.191766  
Wilson High School             81.296540          67.455103  
Ford High School               82.219788          67.469880

In [114]:
top_schools.to_csv("Output/bottom_schools.csv", encoding="utf-8", index=False, header = True)

## Maths Scores by Year

In [175]:
# Use the code provided to separate the data by grade
year_nine = school_data_complete[(school_data_complete["year"] == 9)]
year_ten = school_data_complete[(school_data_complete["year"] == 10)]
year_eleven = school_data_complete[(school_data_complete["year"] == 11)]
year_twelve = school_data_complete[(school_data_complete["year"] == 12)]

# Group by `school_name` and take the mean of the the `maths_score` column for each.
year_nine_scores = year_nine.groupby(["school_name"])
year_nine_scores = year_nine_scores["maths_score"].mean()
year_ten_scores = year_ten.groupby(["school_name"])
year_ten_scores = year_ten_scores["maths_score"].mean()
year_eleven_scores = year_eleven.groupby(["school_name"])
year_eleven_scores = year_eleven_scores["maths_score"].mean()
year_twelve_scores = year_twelve.groupby(["school_name"])
year_twelve_scores = year_twelve_scores["maths_score"].mean()
# Combine each of the scores above into single DataFrame called `maths_scores_by_grade`
maths_scores_by_year = pd.DataFrame({
        "Year 9": year_nine_scores,
        "Year 10": year_ten_scores,
        "Year 11": year_eleven_scores,
        "Year 12": year_twelve_scores
})

# Minor data wrangling
maths_scores_by_year = maths_scores_by_year[["Year 9", "Year 10", "Year 11", "Year 12"]]
maths_scores_by_year.index.name = None

# Display the DataFrame
maths_scores_by_year

Year 9    Year 10    Year 11    Year 12
Bailey High School     72.493827  71.897498  72.374900  72.675097
Cabrera High School    72.321970  72.437768  71.008299  70.604712
Figueroa High School   68.477804  68.331586  68.811001  69.325282
Ford High School       69.021609  69.387006  69.248862  68.617811
Griffin High School    72.789731  71.093596  71.692521  71.469178
Hernandez High School  68.586831  68.867156  69.154412  68.985075
Holden High School     70.543307  75.105263  71.640777  73.409639
Huang High School      69.081754  68.533246  69.431345  68.639316
Johnson High School    69.469286  67.990220  68.637730  69.287393
Pena High School       71.996364  72.396000  72.523438  71.187845
Rodriguez High School  71.940722  71.779808  72.364811  72.154626
Shelton High School    72.932075  72.506696  70.097087  72.331536
Thomas High School     69.234273  70.057007  69.657831  69.369822
Wilson High School     69.212361  69.455446  68.378965  69.787472
Wright High School     71.741176  72.179012  73.275862  70.848238

## Reading Score by Year

In [176]:
# Use the code provided to separate the data by grade
year_nine = school_data_complete[(school_data_complete["year"] == 9)]
year_ten = school_data_complete[(school_data_complete["year"] == 10)]
year_eleven = school_data_complete[(school_data_complete["year"] == 11)]
year_twelve = school_data_complete[(school_data_complete["year"] == 12)]

# Group by `school_name` and take the mean of the the `reading_score` column for each.
year_nine_scores = year_nine.groupby(["school_name"])
year_nine_scores = year_nine_scores["reading_score"].mean()
year_ten_scores = year_ten.groupby(["school_name"])
year_ten_scores = year_ten_scores["reading_score"].mean()
year_eleven_scores = year_eleven.groupby(["school_name"])
year_eleven_scores = year_eleven_scores["reading_score"].mean()
year_twelve_scores = year_twelve.groupby(["school_name"])
year_twelve_scores = year_twelve_scores["reading_score"].mean()
# Combine each of the scores above into single DataFrame called `reading_scores_by_grade`
reading_scores_by_year = pd.DataFrame({
        "Year 9": year_nine_scores,
        "Year 10": year_ten_scores,
        "Year 11": year_eleven_scores,
        "Year 12": year_twelve_scores
})

# Minor data wrangling
reading_scores_by_year = reading_scores_by_year[["Year 9", "Year 10", "Year 11", "Year 12"]]
reading_scores_by_year.index.name = None

# Display the DataFrame
reading_scores_by_year

Year 9    Year 10    Year 11    Year 12
Bailey High School     70.901920  70.848265  70.317346  72.195525
Cabrera High School    71.172348  71.328326  71.201245  71.856021
Figueroa High School   70.261682  67.677588  69.152327  69.082126
Ford High School       69.615846  68.988701  70.735964  68.849722
Griffin High School    72.026895  70.746305  72.385042  69.434932
Hernandez High School  68.477569  70.621842  68.418199  69.244136
Holden High School     71.598425  71.096491  73.310680  70.481928
Huang High School      68.670616  69.516297  68.740638  68.671795
Johnson High School    68.719286  69.295029  69.969115  67.992521
Pena High School       70.949091  72.324000  71.703125  71.513812
Rodriguez High School  70.902921  70.137500  71.424453  71.414449
Shelton High School    70.715094  69.879464  71.150485  69.070081
Thomas High School     69.672451  69.741093  70.749398  68.730769
Wilson High School     68.683043  68.412541  68.796327  69.888143
Wright High School     71.823529  71.386831  70.475862  69.821138

## Scores by School Spending

In [182]:
# Establish the bins 
spending_bins = [0, 585, 630, 645, 680]
group_names = ["<$585", "$585-630", "$630-645", "$645-680"]

# Create a copy of the school summary since it has the "Per Student Budget" 
school_spending_df = per_school_summary.copy()

# Use `pd.cut` to categorize spending based on the bins.
school_spending_df['Per Student Budget'] = school_spending_df['Per Student Budget'].replace('[\$,]', '', regex=True).astype(float)

school_spending_df["Spending Ranges (Per Student)"] = pd.cut(per_school_capita, spending_bins, labels = group_names)
school_spending_df

School Type  Total Students  Total School Budget  \
school_name                                                               
Bailey High School      Government            4976              3124928   
Cabrera High School    Independent            1858              1081356   
Figueroa High School    Government            2949              1884411   
Ford High School        Government            2739              1763916   
Griffin High School    Independent            1468               917500   
Hernandez High School   Government            4635              3022020   
Holden High School     Independent             427               248087   
Huang High School       Government            2917              1910635   
Johnson High School     Government            4761              3094650   
Pena High School       Independent             962               585858   
Rodriguez High School   Government            3999              2547363   
Shelton High School    Independent            1761              1056600   
Thomas High School     Independent            1635              1043130   
Wilson High School     Independent            2283              1319574   
Wright High School     Independent            1800              1049400   

                       Per Student Budget  Average Maths Score  \
school_name                                                      
Bailey High School                  628.0            72.352894   
Cabrera High School                 582.0            71.657158   
Figueroa High School                639.0            68.698542   
Ford High School                    644.0            69.091274   
Griffin High School                 625.0            71.788147   
Hernandez High School               652.0            68.874865   
Holden High School                  581.0            72.583138   
Huang High School                   655.0            68.935207   
Johnson High School                 650.0            68.843100   
Pena High School                    609.0            72.088358   
Rodriguez High School               637.0            72.047762   
Shelton High School                 600.0            72.034072   
Thomas High School                  638.0            69.581651   
Wilson High School                  578.0            69.170828   
Wright High School                  583.0            72.047222   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  \
school_name                                                   
Bailey High School             87.379421          80.084405   
Cabrera High School            89.074273          80.785791   
Figueroa High School           82.807731          67.650051   
Ford High School               82.219788          67.469880   
Griffin High School            88.487738          81.335150   
Hernandez High School          81.877023          66.364617   
Holden High Sc

In [147]:
#  Calculate averages for the desired columns. 
spending_maths_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Maths Score"].mean()
spending_reading_scores = school_spending_df.groupby(["Spending Ranges (Per Student)"])["Average Reading Score"].mean()
spending_passing_maths = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Maths"].mean()
spending_passing_reading = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Passing Reading"].mean()
overall_passing_spending = school_spending_df.groupby(["Spending Ranges (Per Student)"])["% Overall Passing"].mean()

# import warnings filter
from warnings import simplefilter
# ignore all future warnings
simplefilter(action='ignore', category=FutureWarning)

In [148]:
# Assemble into DataFrame
spending_summary = pd.DataFrame({
          "Average Maths Score" : spending_maths_scores,
          "Average Reading Score": spending_reading_scores,
          "% Passing Maths": spending_passing_maths,
          "% Passing Reading": spending_passing_reading,
          "% Overall Passing": overall_passing_spending
})
# Clean up output
spending_summary["Average Maths Score"] = spending_summary["Average Maths Score"].map("{:.2f}".format)
spending_summary["Average Reading Score"] = spending_summary["Average Reading Score"].map("{:.2f}".format)
spending_summary["% Passing Maths"] = spending_summary["% Passing Maths"].map("{:.2f}".format)
spending_summary["% Passing Reading"] = spending_summary["% Passing Reading"].map("{:.2f}".format)
spending_summary["% Overall Passing"] = spending_summary["% Overall Passing"].map("{:.2f}".format)

# Display results
spending_summary

Average Maths Score Average Reading Score  \
Spending Ranges (Per Student)                                             
<$585                                       71.36                 70.72   
$585-630                                    72.07                 71.03   
$630-645                                    69.85                 69.84   
$645-680                                    68.88                 69.05   

                              % Passing Maths % Passing Reading  \
Spending Ranges (Per Student)                                     
<$585                                   88.84             86.39   
$585-630                                91.52             87.29   
$630-645                                84.69             83.76   
$645-680                                81.57             81.77   

                              % Overall Passing  
Spending Ranges (Per Student)                    
<$585                                     76.72  
$585-630                                  79.88  
$630-645                                  71.00  
$645-680                                  66.76

## Scores by School Size

In [154]:
# Establish the bins.
size_bins = [0, 1000, 2000, 5000]
labels = ["Small (<1000)", "Medium (1000-2000)", "Large (2000-5000)"]

# Categorize the spending based on the bins
per_school_summary["School Size"] = pd.cut(per_school_summary["Total Students"], size_bins, labels=labels)

per_school_summary

School Type  Total Students  Total School Budget  \
school_name                                                               
Bailey High School      Government            4976              3124928   
Cabrera High School    Independent            1858              1081356   
Figueroa High School    Government            2949              1884411   
Ford High School        Government            2739              1763916   
Griffin High School    Independent            1468               917500   
Hernandez High School   Government            4635              3022020   
Holden High School     Independent             427               248087   
Huang High School       Government            2917              1910635   
Johnson High School     Government            4761              3094650   
Pena High School       Independent             962               585858   
Rodriguez High School   Government            3999              2547363   
Shelton High School    Independent            1761              1056600   
Thomas High School     Independent            1635              1043130   
Wilson High School     Independent            2283              1319574   
Wright High School     Independent            1800              1049400   

                       Per Student Budget  Average Maths Score  \
school_name                                                      
Bailey High School                  628.0            72.352894   
Cabrera High School                 582.0            71.657158   
Figueroa High School                639.0            68.698542   
Ford High School                    644.0            69.091274   
Griffin High School                 625.0            71.788147   
Hernandez High School               652.0            68.874865   
Holden High School                  581.0            72.583138   
Huang High School                   655.0            68.935207   
Johnson High School                 650.0            68.843100   
Pena High School                    609.0            72.088358   
Rodriguez High School               637.0            72.047762   
Shelton High School                 600.0            72.034072   
Thomas High School                  638.0            69.581651   
Wilson High School                  578.0            69.170828   
Wright High School                  583.0            72.047222   

                       Average Reading Score  % Passing Maths  \
school_name                                                     
Bailey High School                 71.008842        91.639871   
Cabrera High School                71.359526        90.850377   
Figueroa High School               69.077993        81.654798   
Ford High School                   69.572472        82.438846   
Griffin High School                71.245232        91.212534   
Hernandez High School              69.186408        80.949299   
Holden High School                 71.660422        89.929742   
Huang High School                  68.910525        81.693521   
Johnson High School                69.039277        82.062592   
Pena High School                   71.613306        91.683992   
Rodriguez High School              70.935984        90.797699   
Shelton High School                70.257808        91.538898   
Thomas High School                 69.768807        83.853211   
Wilson High School                 68.876916        82.785808   
Wright High School                 70.969444        91.777778   

                       % Passing Reading  % Overall Passing  \
school_name                                                   
Bailey High School             87.379421          80.084405   
Cabrera High School            89.074273          80.785791   
Figueroa High School           82.807731          67.650051   
Ford High School               82.219788          67.469880   
Griffin High School            88.487738          81.335150   
Hernandez High School          81.877023          66.364617   
Holden High Sc

In [149]:
# Create new series using groupby for:
# Type | Average Maths Score | Average Reading Score | % Passing Maths | % Passing Reading | % Overall Passing

size_maths_scores = per_school_summary.groupby(["School Size"])["Average Maths Score"].mean()
size_reading_scores = per_school_summary.groupby(["School Size"])["Average Reading Score"].mean()
size_passing_maths = per_school_summary.groupby(["School Size"])["% Passing Maths"].mean()
size_passing_reading = per_school_summary.groupby(["School Size"])["% Passing Reading"].mean()
size_overall_passing = per_school_summary.groupby(["School Size"])["% Overall Passing"].mean()

In [153]:
# Create a DataFrame called `size_summary` that breaks down school performance based on school size (small, medium, or large).
# Use the scores above to create a new DataFrame called `size_summary`
size_summary = pd.DataFrame({
    "Average Maths Score": size_maths_scores,
    "Average Reading Score": size_reading_scores,
    "% Passing Maths": size_passing_maths,
    "% Passing Reading": size_passing_reading,
    "% Overall Passing": size_overall_passing
})

# Display results
size_summary

Average Maths Score  Average Reading Score  \
School Size                                                      
Small (<1000)                 72.335748              71.636864   
Medium (1000-2000)            71.421650              70.720164   
Large (2000-5000)             69.751809              69.576052   

                    % Passing Maths  % Passing Reading  % Overall Passing  
School Size                                                                
Small (<1000)             90.806867          87.557513          79.066348  
Medium (1000-2000)        89.846560          86.714149          78.039785  
Large (2000-5000)         84.252804          83.301185          70.293507

## Scores by School Type

In [101]:
# Create new series using groupby for:
# Type | Average Maths Score | Average Reading Score | % Passing Maths | % Passing Reading | % Overall Passing

type_maths_scores = per_school_summary.groupby(["School Type"])["Average Maths Score"].mean()
type_reading_scores = per_school_summary.groupby(["School Type"])["Average Reading Score"].mean()
type_passing_maths = per_school_summary.groupby(["School Type"])["% Passing Maths"].mean()
type_passing_reading = per_school_summary.groupby(["School Type"])["% Passing Reading"].mean()
type_overall_passing = per_school_summary.groupby(["School Type"])["% Overall Passing"].mean()

In [186]:
type_summary = per_school_summary.groupby("School Type")
type_summary = type_summary[["Average Maths Score", "Average Reading Score", "% Passing Maths", 
                             "% Passing Reading", "% Overall Passing"]]
type_summary.mean()

Average Maths Score  Average Reading Score  % Passing Maths  \
School Type                                                                
Government             69.834806              69.675929        84.462375   
Independent            71.368822              70.718933        89.204043   

             % Passing Reading  % Overall Passing  
School Type                                        
Government           83.587562          70.698993  
Independent          86.247789          76.973340

# Data Analysis Report
The district summary DataFrame show that throughtout 15 schools with 39,170 total student and over 24.5 millon dollars, students are more successful at reading than math. From the collected data we can conclude that higher spending (per student), does not result in higher test scores. It can also be concluded that small (<1000) schools have better overall passing then large (2000 - 5000) schools. Summariing all the above points, we can conclude the most effective school with average costs is griffin Hight school. (% Overall passing 81.335; budget per student $625)

Two Conclusions:
1. Independent Schools are better placed to achieve higher passing grades on average.(please check the data (type_summary.mean())
2. The results of scores by School Spending and the scores by School Size, the top performing schools have some of the lowest number of total students. Smaller schools with smaller per student budgets perform better in comparison with bigger schools with greater per student budgets.
